statistics/dml_update.go

```diff
@@ -0,0 +1,271 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package statistics
+
+import (
+	"github.com/cznic/mathutil"
+	"github.com/pingcap/tidb/model"
+	"github.com/pingcap/tidb/sessionctx/stmtctx"
+	"github.com/pingcap/tidb/terror"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/codec"
+)
+
+// TableDelta stands for the changed stats for one table.
+// Note that we only update stats for primary key and indices.
+type TableDelta struct {
+	Delta     int64 // Delta is the delta count for a table.
+	Count     int64 // Count is the modify count for a table.
+	PKID      int64
+	PKDelta   *HistDelta
+	IdxDeltas map[int64]*HistDelta
+}
+
+// HistDelta stands for the changed stats for one histogram.
+type HistDelta struct {
+	HistVersion uint64 // HistVersion is the version of the histogram when the delta info is collected.
+	BktDeltas   map[int]*BktDelta
+}
+
+// BktDelta stands for the delta info for a bucket.
+type BktDelta struct {
+	Count int64
+	Bound *types.Datum // Bound is the new lower bound of the bucket. For the last bucket, it is also used to store the new upper bound of the bucket.
+}
+
+// Update updates the table delta according to the changed row.
+func (delta *TableDelta) Update(sc *stmtctx.StatementContext, h *Handle, tableInfo *model.TableInfo, row []types.Datum, deltaCount, count int64) {
+	delta.Delta += deltaCount
+	delta.Count += count
+	table := h.GetTableStats(tableInfo.ID)
+	// Update the primary key's histogram.
+	if tableInfo.PKIsHandle {
+		pk := tableInfo.GetPkColInfo()
+		c, ok := table.Columns[pk.ID]
+		if ok && pk.State == model.StatePublic && pk.Offset >= 0 && pk.Offset < len(row) {
```

Comment:
When the offset of pk will be out of range ?

Reply:
Do not know, just to make sure that it won't cause panic.